### ISSUE TITLE
Not shaking the tree...

### ISSUE THREAD

**Issue by Maze-fr** [Myrmod/svelte-babylon#13]
Hello,

Very simple example :
1. Create a new page in your routes
2. Create a Canvas in that page :
```html
<script lang="ts">
    import { Canvas } from 'svelte-babylon';
</script>

<Canvas />
```
3. Then a simple `npm run build`
4. Check the size of the JS associated to that new page in `build/_app/pages/` : it is 5Mb !

**Edit :** maybe... those 2 lines in `Canvas/index.svelte` may be the problem :
```JavaScript
import * as BABYLON from 'babylonjs'
import * as cannon from 'cannon'
```

**Comment by Maze-fr** [comment]
I did some tests...
By declaring just the needed classes, it drops to 4.3Mb.
If I remove `cannon`, it drops to 4.1Mb : maybe `cannon` should be specified during dev if needed.

20% smaller !
But I think it is still way too excessive for just rendering a canvas...

**Comment by Maze-fr** [comment]
Using separated packages may enable the tree shaking, like described here :
https://doc.babylonjs.com/divingDeeper/developWithBjs/treeShaking

**Comment by Myrmod** [comment]
Hi Maze,

thank you for your feedback. The size you mentioned is definitely far too large.
I just put `cannon` into a dynamic import, this kinda removes it from the main bundle in a basic vite application eg. `npm init vite`.

I tried using the following basic babylonjs implementation:

```svelte
<script lang="ts">
  import { Engine, Scene } from "babylonjs";
  import { onMount } from "svelte";

  let canvas: HTMLCanvasElement;

  onMount(() => {
    if (canvas) {
      const engine = new Engine(canvas);
      const scene = new Scene(engine);

      scene.render();
    }
  });
</script>

<canvas bind:this={canvas} />
```
Which using `npm run build` results in 
```sh
vite v2.9.4 building for production...
✓ 9 modules transformed.
dist/index.html                 0.37 KiB
dist/assets/index.2545e538.js   3999.85 KiB / gzip: 874.23 KiB
```

I will try to look more into this over the weekend. I am sadly quite busy with my job.

I have some questions though:
```
By declaring just the needed classes, it drops to 4.3Mb.
```
What exactly did you change?

```
I think maybe the problem is that ALL properties are mandatory ALMOST EVERYWHERE in the TS types... so you grab one type definition, you grab them all...
``` 
Could you elaborate with an example please?

**Comment by Maze-fr** [comment]
I copied / pasted your Canvas component folder in my project, and I replaced the * import by :
```JavaScript
import BABYLON from 'babylonjs';
const { Vector3, Color4, Engine, Scene, CannonJSPlugin } = BABYLON;
```
And then I just used your Canvas component without any parameter.
I also tried with your test code : same result, as it did for you.

I don't know if it drains 4Mb because of the strong typing (almost no `myProp?: WhateverObject`, for the physics for example), but as it shrinks a bit when I limited the types imports, I think the way it is bundled (everything with a single `BABYLON` object module) is probably not the reason.
From what I read, using separated packages should reduce the size a lot.

We can also wait for Babylon 6 in one year :-P
And about that, I fear one thing : Babylon wants to always be retro-compatible. So it's difficult to deconstruct and rebuild bad / "old way" JS coding.

**Comment by Myrmod** [comment]
> 

I will first try to follow https://doc.babylonjs.com/divingDeeper/developWithBjs/treeShaking. This should, as far as I understand, reduce the bundle size dramatically.

Babylonjs wants to remove IE support in v6, so some improvement will come later down the line anyway.

**Comment by Myrmod** [comment]
I started changing the imports and replaced all `babylonjs` imports with `@babylonjs/*` packages. This helpsjust a little, so as far as I understand I need to specify the imports like this
```js
import {
    AmmoJSPlugin,
    CannonJSPlugin,
    Color4,
    Engine,
    EngineOptions,
    OimoJSPlugin,
    Scene,
    Vector3,
  } from '@babylonjs/core'
```
becomes
````js
 import { Engine } from '@babylonjs/core/Engines/engine'
 import type { EngineOptions } from '@babylonjs/core/Engines/thinEngine'
 import { Color4 } from '@babylonjs/core/Maths/math.color'
 import { Vector3 } from '@babylonjs/core/Maths/math.vector'
 import type { AmmoJSPlugin } from '@babylonjs/core/Physics/Plugins/ammoJSPlugin'
 import { CannonJSPlugin } from '@babylonjs/core/Physics/Plugins/cannonJSPlugin'
 import type { OimoJSPlugin } from '@babylonjs/core/Physics/Plugins/oimoJSPlugin'
 import { Scene } from '@babylonjs/core/scene'
```

However, I didn't find any changes in bundle size doing this. I sit at 4.5mb. I guess I am missing something.

branch: https://github.com/Myrmod/svelte-babylon/tree/import-refactor

**Comment by Maze-fr** [comment]
Are you talking about your lib complete bundle or the size an app with just a canvas ?

**Comment by Myrmod** [comment]
The app size with just the canvas. Haven't tested it with the BabylonJS only approach yet though. So I can't compare it

**Comment by Maze-fr** [comment]
Try this... I don't say anything, I don't want to spoil...
```html
<script lang="ts">
	import { Engine } from '@babylonjs/core/Engines/engine.js';
	import { Scene } from '@babylonjs/core/scene.js';
	import { onMount } from 'svelte';

	let canvas: HTMLCanvasElement;

	onMount(() => {
		if (!!canvas) {
			const engine = new Engine(canvas);
			const scene = new Scene(engine);

			scene.render();
		}
	});
</script>

<canvas bind:this={canvas} />
```

**Comment by Myrmod** [comment]
> Try this... I don't say anything, I don't want to spoil...
> 
> ```
> <script lang="ts">
> 	import { Engine } from '@babylonjs/core/Engines/engine.js';
> 	import { Scene } from '@babylonjs/core/scene.js';
> 	import { onMount } from 'svelte';
> 
> 	let canvas: HTMLCanvasElement;
> 
> 	onMount(() => {
> 		if (!!canvas) {
> 			const engine = new Engine(canvas);
> 			const scene = new Scene(engine);
> 
> 			scene.render();
> 		}
> 	});
> </script>
> 
> <canvas bind:this={canvas} />
> ```

I got 
```
➜  bundle-size npm run build

> bundle-size@0.0.0 build
> vite build

vite v2.9.5 building for production...
✓ 102 modules transformed.
dist/index.html                 0.37 KiB
dist/assets/index.18d0dc87.js   453.03 KiB / gzip: 101.31 KiB
```

It's 1/10 of when including `svelte-babylon`. So I assume, that I made some mistake, so I can solve this :)

Did you have something else?

**Comment by Myrmod** [comment]
using

```svelte
<script lang="ts">
  import Canvas from 'svelte-babylon/components/Canvas/index.svelte'
</script>

<Canvas />
```
I get
```
vite v2.9.5 building for production...
✓ 141 modules transformed.
dist/index.html                  0.43 KiB
dist/assets/index.590e3bea.css   0.10 KiB / gzip: 0.10 KiB
dist/assets/cannon.d39c7d5e.js   130.70 KiB / gzip: 37.14 KiB
dist/assets/index.050e3e6f.js    757.32 KiB / gzip: 170.87 KiB
```

which is a lot better

**Comment by Maze-fr** [comment]
ThreeJS minified is 623Kb, and "tree shackable" by the special minifier at 354Kb.
So 453Kb is a big and surprising improvement compared to 4Mb, but I'm still disappointed to see something so big just for an empty canvas.
I'm curious to see how it will grow depending on the examples you made.

From the talks on the Babylon forum, the 6th version coming next year will be even more "tree shackable" and keeping compatibility, so just by upgrading, it will shrink :-)
So your lib is an investment for the future...

For your canvas (and other components, I guess), I wonder if it wouldn't be better to just initialize a minimum canvas, and passing "this" to sub components using the context, to complete the configuration "atomically".
That way, a lot of things you are doing in your canvas would be delegated to the subcomponents, and so a lot of code only when needed.

**Comment by Myrmod** [comment]
> For your canvas (and other components, I guess), I wonder if it wouldn't be better to just initialize a minimum canvas, and passing "this" to sub components using the context, to complete the configuration "atomically". That way, a lot of things you are doing in your canvas would be delegated to the subcomponents, and so a lot of code only when needed.

Sounds interesting. If you could provide an example on what you mean exactly I can take a deeper looker after making the changes for this issues.

You already helped me a lot, thank you :)

**Comment by Maze-fr** [comment]
So... I made this quick PoC, and I didn't get any error when launching it, so I guess it works. But there is probably a way to do it more efficiently.
I used stores in the context because I get the canvas only when its HTML get rendered, so I had to wait for it to be initialized.
```html
<!-- Canvas.svelte -->
<script lang="ts">
	import { setContext } from 'svelte';
	import { writable } from 'svelte/store';

	let canvas: HTMLCanvasElement = undefined;
	setContext('canvas', writable(canvas));
</script>

<canvas bind:this={canvas} />
<slot />
```
```html
<!-- Engine.svelte -->
<script lang="ts">
	import { getContext, setContext } from 'svelte';
	import { Engine } from '@babylonjs/core/Engines/engine.js';
	import { writable, type Writable } from 'svelte/store';

	let engine: Writable<Engine> = writable(undefined);
	let canvas: Writable<HTMLCanvasElement> = getContext('canvas');
	setContext('engine', engine);

	$: if (!!$canvas) {
		$engine = new Engine($canvas);
	}
</script>

<slot />
```
```html
<!-- Scene.svelte -->
<script lang="ts">
	import { getContext, onMount, setContext } from 'svelte';
	import { Scene } from '@babylonjs/core/scene.js';
	import type { Engine } from '@babylonjs/core/Engines/engine';
	import { writable, type Writable } from 'svelte/store';

	let scene: Writable<Scene> = writable(undefined);
	let engine: Writable<Engine> = getContext('engine');
	setContext('scene', scene);

	$: if (!!$engine) {
		$scene = new Scene($engine);
		$scene.render();
	}
</script>

<slot />
```
```html
<!-- index.svelte -->
<script lang="ts">
	import Canvas from '$lib/Canvas.svelte';
	import Engine from '$lib/Engine.svelte';
	import Scene from '$lib/Scene.svelte';
</script>

<Canvas>
	<Engine>
		<Scene />
	</Engine>
</Canvas>
```

**Comment by Myrmod** [comment]
I just released version 0.14.0. Eg for the Box example I am now down to
```
vite v2.9.5 building for production...
✓ 266 modules transformed.
dist/index.html                  0.43 KiB
dist/assets/index.590e3bea.css   0.10 KiB / gzip: 0.10 KiB
dist/assets/cannon.d39c7d5e.js   130.70 KiB / gzip: 37.14 KiB
dist/assets/index.794bda38.js    1208.63 KiB / gzip: 279.17 KiB
```

So we save round about 2.7mb. Which is nice. For the gzip I think the bundle size is totally fine for now.



> So... I made this quick PoC, and I didn't get any error when launching it, so I guess it works. But there is probably a way to do it more efficiently. I used stores in the context because I get the canvas only when its HTML get rendered, so I had to wait for it to be initialized.
> 
> ```
> 
> <script lang="ts">
> 	import { setContext } from 'svelte';
> 	import { writable } from 'svelte/store';
> 
> 	let canvas: HTMLCanvasElement = undefined;
> 	setContext('canvas', writable(canvas));
> </script>
> 
> <canvas bind:this={canvas} />
> <slot />
> ```
> 
> ```
> 
> <script lang="ts">
> 	import { getContext, setContext } from 'svelte';
> 	import { Engine } from '@babylonjs/core/Engines/engine.js';
> 	import { writable, type Writable } from 'svelte/store';
> 
> 	let engine: Writable<Engine> = writable(undefined);
> 	let canvas: Writable<HTMLCanvasElement> = getContext('canvas');
> 	setContext('engine', engine);
> 
> 	$: if (!!$canvas) {
> 		$engine = new Engine($canvas);
> 	}
> </script>
> 
> <slot />
> ```
> 
> ```
> 
> <script lang="ts">
> 	import { getContext, onMount, setContext } from 'svelte';
> 	import { Scene } from '@babylonjs/core/scene.js';
> 	import type { Engine } from '@babylonjs/core/Engines/engine';
> 	import { writable, type Writable } from 'svelte/store';
> 
> 	let scene: Writable<Scene> = writable(undefined);
> 	let engine: Writable<Engine> = getContext('engine');
> 	setContext('scene', scene);
> 
> 	$: if (!!$engine) {
> 		$scene = new Scene($engine);
> 		$scene.render();
> 	}
> </script>
> 
> <slot />
> ```
> 
> ```
> 
> <script lang="ts">
> 	import Canvas from '$lib/Canvas.svelte';
> 	import Engine from '$lib/Engine.svelte';
> 	import Scene from '$lib/Scene.svelte';
> </script>
> 
> <Canvas>
> 	<Engine>
> 		<Scene />
> 	</Engine>
> </Canvas>
> ```
> 
> **Edit :** 500Kb for the full build of the PoC project.

For your suggestion I don't think this would improve the developer experience because a developer has to differentiate between `Canvas`, `Engine` and `Scene`.  You cannot use a canvas for multiple scenes, but, you could perhaps use an engine for multiple canvases and scenes. I haven't looked into this that much yet, and don't plan to in the near future, since I think that there're more important features still missing. Like the bundle size which we addressed in this issue.

If you find a good use case for splitting the Canvas into Canvas, Engine and Scene, feel free to open a new issue, where we can discuss how to proceed.

Thank you again for your help!
